Add access rate limit

truongdd03 opened the pull request:
# Description

This PR uses [express-rate-limit](https://www.npmjs.com/package/express-rate-limit) module to set two access limits. The first one (200 requests/ 5 seconds) limits the rate of overall requests to OED. At the moment, the current limit is high due to issue #681. After resolving it, testings are needed to adjust the rate limit. The second limit is specifically for raw export (5 requests/ 5 seconds).

Fixes #266 

## Type of change

- [x] Note merging this changes the node modules
- [ ] Note merging this changes the database configuration.
- [x] This change requires a documentation update

## Checklist

- [x] I have followed the [OED pull request](https://openenergydashboard.github.io/developer/pr.html) ideas
- [x] I have removed text in ( ) from the issue request

## Limitations

Note that the rate limit may make future automatic tests return the value of 429 (too many requests). In that case, the limiters need to be increased.

truongdd03 committed:
install express-rate-limit

truongdd03 committed:
Add rate limit

truongdd03 committed:
Reformat app.js

truongdd03 committed:
Merge branch 'OpenEnergyDashboard:development' into rate_limiting

truongdd03 committed:
Replace package.json and package-lock.json

huss commented on src/server/app.js:
```diff
@@ -30,7 +31,11 @@ const logs = require('./routes/logs');
 const csv = require('./routes/csv');
 
-const app = express();
+const limiter = new rateLimit({
```
Nice to add a comment about the general intent of this code (to limit the rate of overall requests to OED) and what the two values mean.

huss committed:
rate limiting test script

I took the basic script from @truongdd03 and made it a little more
general. I felt it was good to have it in the testing directory so
others would have access to it. Note this is currently only manual
testing but automating at some point might be good, esp. if we
limit specific routes.

huss commented:
This is good work and does basic rate limiting. Testing indicates it works as expected. I generalized the test script written by @truongdd03 and put it in the testing directory for future use.

I have two thoughts on this PR:

1. Why is the overall limit set at 300/minute? It seems like it may be fine but was there testings to see if this would work for sites? Can a single user request may count as multiple requests toward the limit if it makes api calls internally?Just want to know for the record.
2. Do people know if changing the time frame for the reset (currently 1 minute) would help with accidentally hitting it with valid requests at a site?
3. I do think the overall limit cannot be too low as we want log api requests to succeed so log messages are not lost.
2. In the long-term, we might want to have a default value and allow the admin to modify that value. That is beyond this PR but just noting.
2. An overall limit seems good but I think specific routes should probably also have limits. For example, asking for the raw data is going to be much more draining on resources than asking for bar graphic data. The default size to download without a login is 5MB so at 300 requests/min that would be 1.5 GB/min which would certainly tax the system. We don't expect a lot of these requests so maybe limiting to 5-10 per minute might be better. I don't see any other obvious routes that doe not require a login that are an issue like this.

truongdd03 committed:
Add comments

truongdd03 committed:
Increase rate limit and add raw export limit

truongdd03 committed:
Adjust limits

truongdd03 committed:
Fix raw export limit

huss committed:
fix up rate limit testing script

- add test for new raw limit.
- remove extra echos and commented out code.

huss commented:
Good work. I appreciate the effort to test and get all the details working. This rate limits overall requests (allowing a lot) and raw download (only a few). I updated the test script a little to deal with items. It all tests out fine so ready to merge.
